### ISSUE TITLE
Running a cmd.exe script deselects one file

### ISSUE THREAD

**Issue by tukanos** [doublecmd/doublecmd#324]
I have a workflow where I select files do one or multiple transformations.  

I just found out that if I run a command via toolbar the first file gets deselected, which ruins the workflow for multiple transformations on the selected files.

I command in the toolbar is (removing all metadata via exiftool):

```
Command: cmd
Parameters: /C exiftool -all= %ps & pause
```

Here is a screenshot for the bug (notice how the file Copy (1) IMG_20211216_094448.jpg gets deselected without any user action):

https://user-images.githubusercontent.com/731154/148955166-4bfb3068-701b-411d-aae4-9dbe83dda566.mp4

**Comment by alexx2000** [comment]
I guess exiftool rename old files and creates new instead. After this operation selection is not valid anymore. Because it is another/new files. It just so matches that the selection only loses the first one.

**Comment by tukanos** [comment]
> I guess exiftool rename old files and creates new instead.

Yes, your guess is right. The _exiftool_ renames old files and create a new ones instead.

> After this operation selection is not valid anymore. 

Why? There exactly the same names as before only with different size now.  Apparently double commander tries to reselect those new files, because it tries to keep the selection, but loses the selection on the first file.

> It just so matches that the selection only loses the first one.

Yes, but losing the first file makes illogical and inconsistent.  Why the others are selected when the first one is not?  That looks like a bug.

In my humble opinion, there are two possible approaches.  

1) Either you keep the whole selection as it was before. This approach I personally prefer as you can perform multiple transformation with your selection.  As an example, I will show the identical operation in Altap Salamander (x64, 4.0) where it works this way:

https://user-images.githubusercontent.com/731154/149119245-e73331d9-7a8a-4072-b20d-5954e1a85f86.mp4



2) You deselect all previously selected files.  As an example, I will show you the identical operation in Total Commander (x64 10.00):

https://user-images.githubusercontent.com/731154/149119264-0c41a3e6-b034-4184-89c8-f1efbc0c4f4a.mp4

I'm in middle of moving to double commander and so far I'm loving it.  Thank you for it!

**Comment by alexx2000** [comment]
Because file disappear first and some later created new. But DC don't react on all updates (to reduce the processor load). It refreshes panel at most each 500ms. So it doesn't see that all files was renamed and replaced by new. It see that first file was rename/create, for other files it see that only size/time was updated.

So I can implement only variant 2.

**Comment by tukanos** [comment]
Thank you for your explanation.

> Because file disappear first and some later created new. But DC don't react on all updates (to reduce the processor load). It refreshes panel at most each 500ms. So it doesn't see that all files was renamed and replaced by new. It see that first file was renamed/created, for other files it see that only size/time was updated.

I see.   The DC misses the first file selection update because it does not fit the 500ms limit.   Is that 500ms limit hardcoded or it can vary based on operation/user?  Would be possible to allow user to force DC to react on all updates only for this occasion (running a command or script)? 

> So I can implement only variant 2.

That would be probably better than current inconsistent state.   Would it be possible to store the current selection, run the command/script and try to reselect the files after the operation?

**Comment by alexx2000** [comment]
1. Limit is hardcoded. No, "auto-refresh" is very complicated and fully independent mechanism.
2. You can try to do it using built-in [Lua-script](https://doublecmd.github.io/doc/en/lua.html) and internal commands [cm_SaveSelection](https://doublecmd.github.io/doc/en/cmds.html#cm_SaveSelection) / [cm_RestoreSelection](https://doublecmd.github.io/doc/en/cmds.html#cm_RestoreSelection).

**Comment by tukanos** [comment]
> 1. Limit is hardcoded. No, "auto-refresh" is very complicated and fully independent mechanism.

I see.

> 2. You can try to do it using built-in [Lua-script](https://doublecmd.github.io/doc/en/lua.html) and internal commands [cm_SaveSelection](https://doublecmd.github.io/doc/en/cmds.html#cm_SaveSelection) / [cm_RestoreSelection](https://doublecmd.github.io/doc/en/cmds.html#cm_RestoreSelection).

Great idea, I'll try to solve it like this.

Thank you.

**Comment by tukanos** [comment]
I'm leaving this ticket opened so others could also voice their comments if the total deselection (option 2) should be the proper way to proceed.

**Comment by j2969719** [comment]
this case with simultaneous multiple renaming of selected files is not so common. dont think that only for the sake of it its worth removing selection from files every time external applications and scripts are launched.

**Comment by tukanos** [comment]
> this case with simultaneous multiple renaming of selected files is not so common. dont think that only for the sake of it its worth removing selection from files every time external applications and scripts are launched.

I understand your view.  It all depends on your workflows.  For me personally, this is a common and daily used feature.  This is why I have opened this ticket in the first place.

In this case my concern is the inconsistency.  As I have shown on Altap Salamander and Total Commander the correct approach is to either be able to keep selected all files or deselect all them.  

Think of it this way,  if by any chance you get to a faster/slower computer which will be able to complete 3/1 operation within 500ms limit you don't want to end up with unselected 3 files on faster computer and 1 unselected file on slower.

**Comment by j2969719** [comment]
well yes, but due to the autorefresh feature, they can become unselected only if at a certain point in time an external application, not necessarily launched by the user from doublecmd, performs multiple renaming of the selected files.

It seems to me that the more common case is that the user selects files, passes them to an external application for viewing or editing, and then, possibly, also performs some file operations on them.

**Comment by tukanos** [comment]
> well yes, but due to the autorefresh feature, they can become unselected only if at a certain point in time an external application, not necessarily launched by the user from doublecmd, performs multiple renaming of the selected files.

Only certain point in time, but that is the point. If you have uncertain results you can't be sure that any following fast action (not only renaming files, but any file transformation like resizing pictures, extracting/merging pdf, etc.) will be performed on your selection.  Any file transformation can potentially lead to uncertain result because the original selection can be inconsistent.

> It seems to me that the more common case is that the user selects files, passes them to an external application for viewing or editing, and then, possibly, also performs some file operations on them.

To my understanding if you pass it to external application which performs the changes automatically (e.g. macro), if fast enough, will suffer from same issue.  Of course, if you edit it manually, which I try to avoid as much as possible on repetitive tasks, only then save it DC will have enough time to refresh.

**Comment by j2969719** [comment]
as far as I understand, in order to trigger this issue, its necessary that at some point in one of the autorefresh, the selected file must not exist.

im rather talking about, for example, pre-playing audio files before copying somewhere and similar cases, not about scripting. removing selection in such cases will simply be a headache for the user.

**Comment by TheNicker** [comment]
Is the directory refresh is poll based ~500ms , doesn't it use the system's event based File watcher capabilities ? 
If that is the issue and the reason for loosing data, how hard it might be to integrate such system to double commander ? 

It's a bit itchy seeing the `not fixable` yellow label🙂

**Comment by alexx2000** [comment]
DC use file system watcher, but don't refresh panel on each event (because it will be a high CPU load). DC queue events and process them no more often 500 ms.

**Comment by TheNicker** [comment]
So technically no data gets lost, it is just interpretation and order. 
I'm assuming high CPU load is referred to _immediately handle file system events_, **end-to-end**

Could something like that work  ? 

Process events immediately into a shadow state, then resolve the current truth at the next 500ms update cycle. 

High level flowchart: 

File system events -> Shadow state (real-time, authoritative truth) -> UI tick (500ms) ->  **Visual state snapshot** Render + Reconcile file selection
